### PR TITLE
Bring back Marshal.Read/Write overloads that takes object

### PR DESCRIPTION
**Pull request by yizhang82**
The original desktop implementation is implemented on top of mscoree.ND_R* and ND_W * helpers with AsAny marshalling,which is super odd but I can understand the convenience. I've re-implemented the API on top of AsAny marshaler so that it could work cross-plat without having to introduce new helpers into PAL. I've also introduced a bit of error checking for null case (desktop simply AVs).
I've also added a simple test as well.
@jkotas @shrah PTAL
/cc @danmosemsft @Petermarcu

### CODE REVIEW

**Comment by Petermarcu** [comment]
Great. Thanks @yizhang82 . I agree on the perf and also that throwing an exceptions instead of AVing is a good thing :)

File: tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteObject.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
+﻿using System;
```
Comment (danmosemsft) [inline tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteObject.cs]: Do we have any rules about when we test corelib via coreclr/tests vs. corefx tests? The latter makes it easier to test compat with other platforms and also keeps everything in one place. Of course coreclr is convenient..

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 2]
```diff
@@ -749,16 +785,51 @@ public static unsafe void WriteInt64(IntPtr ptr, int ofs, long val)
...
+        // propagate the value back
+        // It's quite slow and is only there for backcompact
+        // I don't think we should spend time optimizing it
+        // People should really call the IntPtr overload instead
+        //====================================================================
+        private static unsafe void WriteValueSlow<T>(object ptr, int ofs, T val, Action<IntPtr, int, T> writeValueHelper)
+        {
+            // We AV on desktop if passing NULL. So this is technically a breaking change but is an improvement
+            if (ptr == null)
+                throw new ArgumentNullException("ptr");
```
Comment (danmosemsft) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: nit: `throw new ArgumentNullException(nameof(ptr));`

File: src/mscorlib/src/System/StubHelpers.cs [hunk 3]
```diff
@@ -700,6 +700,17 @@ static internal IntPtr CreateNativeNCCEventArgsInstance(int action, object newIt
 
     internal static class MngdNativeArrayMarshaler
     {
+        // Needs to match exactly with MngdNativeArrayMarshaler in ilmarshalers.h
+        internal struct MarshalerState
+        {
+            IntPtr  m_pElementMT;
```
Comment (danmosemsft) [inline src/mscorlib/src/System/StubHelpers.cs]: nit:extra space

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 4]
```diff
@@ -576,6 +577,41 @@ public static long ReadInt64(IntPtr ptr)
...
+        // It's quite slow and can return back dangling pointers
+        // It's only there for backcompact
+        // I don't think we should spend time optimizing it
+        // People should really call the IntPtr overload instead
+        //====================================================================
+        static unsafe T ReadValueSlow<T>(object ptr, int ofs, Func<IntPtr, int, T> readValueHelper)
+        {
+            // We AV on desktop if passing NULL. So this is technically a breaking change but is an improvement
+            if (ptr == null)
+                throw new ArgumentNullException("ptr");
```
Comment (danmosemsft) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: nit: `throw new ArgumentNullException(nameof(ptr));`

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 5]
```diff
@@ -408,7 +409,7 @@ public static void Copy(IntPtr source, IntPtr[] destination, int startIndex, int
         //====================================================================
         public static byte ReadByte([MarshalAs(UnmanagedType.AsAny), In] Object ptr, int ofs)
         {
-            throw new PlatformNotSupportedException(); // https://github.com/dotnet/coreclr/issues/10442
+            return ReadValueSlow(ptr, ofs, (IntPtr nativeHome, int offset) => { return Marshal.ReadByte(nativeHome, offset); });
```
Comment (justinvp) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: Nit: Could be simplified as:

```c#
return ReadValueSlow(ptr, ofs, (nativeHome, offset) => Marshal.ReadByte(nativeHome, offset));
```

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs [hunk 6]
```diff
@@ -576,6 +577,41 @@ public static long ReadInt64(IntPtr ptr)
...
         }
 
+        //====================================================================
+        // Read value from marshaled object (marshaled using AsAny)
+        // It's quite slow and can return back dangling pointers
+        // It's only there for backcompact
+        // I don't think we should spend time optimizing it
+        // People should really call the IntPtr overload instead
+        //====================================================================
+        static unsafe T ReadValueSlow<T>(object ptr, int ofs, Func<IntPtr, int, T> readValueHelper)
```
Comment (justinvp) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: Nit: Specify visibility, e.g. `private static unsafe`

File: tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteObject.cs (hunk 1 above)
Comment (yizhang82) [inline tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteObject.cs]: I don't think we have a policy in place.  In general for the more 'runtime'-y features such as interop it makes sense to put the tests in CoreCLR. Marshal APIs are sort like a gray area that can land either way. My preference is to put them in CoreCLR tests unless it's super clear they belong in CoreFX (such as in System.Runtime.InteropServices.dll, etc). Moving forward, I think we'd want CoreCLR tests to run on multiple runtime configuration as well (they are runtime tests that applies everywhere, after all)

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs (hunk 2 above)
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: That's better. Thanks

File: src/mscorlib/src/System/StubHelpers.cs (hunk 3 above)
Comment (yizhang82) [inline src/mscorlib/src/System/StubHelpers.cs]: Fixed.

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs (hunk 5 above)
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: Thanks. I agree this looks simpler but given that Marshal.ReadByte has different overloads that takes object/IntPtr I'd like to make this a bit more explicit.

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs (hunk 6 above)
Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: fixed.

File: src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs (hunk 5 above)
Comment (justinvp) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: You can keep the parameter types and still remove the unnecessary return statement, e.g.

```c#
return ReadValueSlow(ptr, ofs, (IntPtr nativeHome, int offset) => Marshal.ReadByte(nativeHome, offset));
```

Comment (yizhang82) [inline src/mscorlib/src/System/Runtime/InteropServices/Marshal.cs]: OK. That makes sense. changed

**Pull request merged**